### ISSUE TITLE
Other connected Bluetooth devices disconnect during iOS mirroring

### ISSUE THREAD

**Issue by crouchingtiger** [koush/vysor.io#678]
**Describe the bug**
Other connected BT devices will disconnect during iOS mirroring session.

**To Reproduce**
1. Connect an iOS device with Vysor.
2. Move the mouse / press some keyboard presses on an external keyboard / mouse.
3. Watch your mouse cursor slow down, and eventually see your external BT mouse and keyboard disconnect.
4. Stop / close the Vysor iOS session and the external keyboard / mouse will immediately reconnect.

**Expected behavior**
This might be a limitation of the simultaneous BT connection capabilities of Macs - I have multiple devices connected most of the time. A Magic Keyboard, a Magic Mouse, Bose BT headphones, and there is an Apple Watch and/or Airpods usually nearby (but not all 5 devices are connected at once). As I always use an external keyboard/mouse, I would expect that they are not disconnected due to the BT connection initiated by Vysor / iPhone.

**Desktop (please complete the following information):**
 - OS: macOS 11.1
 - Vysor Version 4.0.7

**iPhone (please complete the following information):**
 - Device: iPhone 11 Pro Max
 - OS: iOS 14.3

**Comment by crouchingtiger** [comment]
I think it's the Bose headphones causing all the interference. I fired up Bluetooth Explorer and watched my connected devices data rates. These headphones can connect to two devices simultaneously, and they were connecting to both the iPhone and Mac handling the the mirroring session. I wasn't listening to anything, but the headphones (perhaps due to the microphone?) are transmitting lots of data constantly. 

I've disconnected the headphones and the interference for Vysor seems to have disappeared almost entirely, and I get instant IO input to the iOS device itself, with the updates reflected in Vysor on the Mac maybe 2-300ms later.

<img width="661" alt="Bildschirmfoto 2021-01-02 um 16 09 34" src="https://user-images.githubusercontent.com/3936238/103460240-08124880-4d15-11eb-8efe-15b368059b3a.png">

**Comment by koush** [comment]
Vysor is only using the BT connection for keyboard+mouse input. The mouse resolution is metered to 1000 HID packets per second, which is very low overall bandwidth.

I have noticed intermittent issues where Vysor's input gets very laggy but have been unable to track it down. I also suspect bluetooth interference, but have not gotten a consistent repro.

**Comment by koush** [comment]
I have some AirPods here that I generally have paired with my computer while using Vysor. I'll try switching to the phone and seeing if things deteriorate.

**Comment by koush** [comment]
I’ve also seen other reports that mac bluetooth gets finicky if there’s slow usb devices on the same usb controller as the bluetooth device (which is internally a usb device).

**Comment by bluepotionmaker** [comment]
I have the similar issues. After connecting to iOS device, my bluetooth mouse is bit laggy and input to the iOS Vysor is unable to track.

**Comment by koush** [comment]
@bluepotionmaker Has this problem persisted? Bluetooth gets sluggish while on the Bluetooth Discovery screens on iOS/mac.

**Comment by koush** [comment]
@bluepotionmaker @crouchingtiger Please let me know your mac model (the new bug report template now asks for this information). This may be specific to the bluetooth hardware in the Mac.

**Comment by bluepotionmaker** [comment]
My mac model is MacBook Pro 16inch, 2019 with BigSur 11.1. 
My iPhone model is iphone6s with iOS 14.3 
Sluggish bluetooth is not persisted but untracked pointer problem always happens. 
I attache the sample video. You can see the inconsistency of the mouse and touch action.

**Comment by koush** [comment]
@bluepotionmaker can you turn on assistive touch? The new version will prompt you to do this. Does seem like your pointer and ios device are getting out of sync, which is a known issue. There’s quite a mouse filter algorithm that tries to keep the ios mouse position in sync with the desktop mouse position. Assistive touch will let you see where the ios mouse is located on screen.

**Comment by bluepotionmaker** [comment]
@koush  Thanks for your quick response. I turned on the assistive touch but it still does not work. 
Should I open the new issue? 

https://user-images.githubusercontent.com/76267121/103496082-3496a380-4e80-11eb-9351-41565101de76.mov

**Comment by koush** [comment]
@bluepotionmaker please open a new issue and attach a bug report. I think I have a fix for this.

**Comment by crouchingtiger** [comment]
It's a MacBookPro16,1 using a Broadcom 4364B3 chip.

After resetting the device config (as suggested in #677), and pausing music that might be playing through my Bose QC 35 headphones, I don't seem to have the issue long-term. If I restart Vysor or the connection, the mouse lag will appear momentarily when Vysor is establishing the BT connection to the iPhone, but after that it seems okay.

I cannot continue to listen to any music through my BT headphones while Vysor is connected (it stutters heavily), but at least my IO seems okay for now.

**Comment by koush** [comment]
Yes, reload vysor will automatically apply the downloaded update if it was ignored.

**Comment by koush** [comment]
> I cannot continue to listen to any music through my BT headphones while Vysor is connected (it stutters heavily), but at least my IO seems okay for now.

For what it's worth, I use AirPods while using Vysor, and it works without any lag. I'm on a hackintosh that is using a OEM Mac  (BCM) wifi/bluetooth too. https://www.amazon.com/gp/product/B00RCKZ3SU/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1

Would be interested to see if it works fine with AirPods on your setup, so I can determine if it is the MacBook or the headphones (likely the headphones).

**Comment by crouchingtiger** [comment]
It's even worse with the AirPods Pro, they don't allow any sound throughput at all, but I can see the data is going to them over BT. I tested without Bluetooth Explorer open first. Also made sure to disconnect them manually from the phone, so that they were only connected to the Mac at the time.

<img width="700" alt="Bildschirmfoto 2021-01-07 um 16 19 59" src="https://user-images.githubusercontent.com/3936238/103909902-87808d00-5104-11eb-854f-92c1f24470c3.png">

**Comment by koush** [comment]
I am using the latest airpod pros and airpod max.
Audio switching works, both are connected, etc. Everything behaves as it should, and works great.

The fact that audio is being sent, but somehow not played, does lead me to believe that Vysor is somehow interfering with outgoing/incoming bluetooth at the connection level (rather than it being simply interference).

Can you attach a bug report? Vysor’s bluetooth service will have a lot of useful info about the connections it observed/accepted, etc.

**Comment by koush** [comment]
Mac 11.2 had a bunch of bluetooth fixes. Does this issue still persist?

<img width="652" alt="image" src="https://user-images.githubusercontent.com/73924/106581433-b3224780-64f7-11eb-94a2-9362f403a0c6.png">

**Comment by crouchingtiger** [comment]
@koush Sorry haven't had a chance to get the report - and also didn't want to update to the 11.2 dev beta due to some existing Xcode project compatibility. I won't be able to update for a few weeks to test 100%. I'm also using Android mirroring most of the time so it's not been a pressing issue for me. Maybe @bluepotionmaker can test?

**Comment by crouchingtiger** [comment]
Finally managed to update to macOS 11.2. There's some BT slow-down of BT mouse / keyboard around the time the session initiates in Vysor, but I'm not seeing total slow-down or disconnection of BT devices *unless* I enable my Bose QC headphones and try to connect. The Mac will then drop BT connections to the Magic Mouse and Keyboard (and fail to connect to the headphones).

**Comment by hixus** [comment]
I can also confirm this. I have Sony wh-1000xm4 which are unsusable when ios screen sahre is on. Also the apple's magic keyboard has huge lag when vysor is on (even if the headphones where not connected).

**Comment by koush** [comment]
> I can also confirm this. I have Sony wh-1000xm4 which are unsusable when ios screen share is on. Also the apple's magic keyboard has huge lag when ios screen share is on (even if the headphones where not connected).

@hixus Can you specify the Mac and iPhone models please?

**Comment by crouchingtiger** [comment]
Just to also add an update - since my last posts I sold my Bose QC 35's a few months ago and now use AirPod Max's as my headphones. The behaviour with them mimics that of the AirPod Pro's previously - that is that any existing music that's playing will stop playing as soon as an iOS Vysor session begins with an iOS device. No issues when mirroring with Android. It doesn't matter what the music source is (Music / Daftcloud / Safari).

macOS is now on v11.3.1, Vysor 4.0.49.

**Comment by koush** [comment]
I just remembered I also have an ancient (2013) MacBook. Let me double check Vysor with some bluetooth accessories on that.

Currently I'm on a hackintosh using a Broadcom bluetooth adapter, and a MacBook Pro M1, with no issues on either. My only bluetooth accessory is also an AirPod Max. I run my mouse/keyboard on Logitech's not-bluetooth adapter (as that's the only way it works in BIOS).

**Comment by hixus** [comment]
> > I can also confirm this. I have Sony wh-1000xm4 which are unsusable when ios screen share is on. Also the apple's magic keyboard has huge lag when ios screen share is on (even if the headphones where not connected).
> 
> @hixus Can you specify the Mac and iPhone models please?

Hi, models are Iphone SE2020 and MacBook Pro (16-inch, 2019). Bluetooth on macbook is Broadcom 4364B3 (firmware v64 c4186). 

I'm not familiar with debugging bluetooth at all but could try if needed and simple enough instructions.

**Comment by hixus** [comment]
related to this would it be possible to disable bluetooth with some setting for now? Don't really need external keyboard + mouse support

**Comment by koush** [comment]
> related to this would it be possible to disable bluetooth with some setting for now? Don't really need external keyboard + mouse support

Yeah, planned to do this in the next release with windows support which often doesn't even have Bluetooth 😅
